README.md
# Pisces
Pisces is a Magic: the Gathering tabletop simulator designed for goldfishing and casual Commander play.

You can [view the Changelog here](CHANGELOG.md).

## Features
- Intuitive UI to easily play cards, change zone, shuffle, and more
- Fast, fluid animation for a more immersive tabletop experience
- Add your own backgrounds and card sleeves (745 x 1040) for a personal touch
- Import decklists or use do a Scryfall search from within Pisces to find the cards or tokens you need
- Automatic token lookup for cards that create them, from that card's right-click menu
- Play with friends on Discord or Spelltable using OBS (v28 or higher) Virtual Camera - and Pisces hides from that view the actual face of cards in hidden zones like your hand or library

## Screenshots
![alt text](https://gamesfreaksa.info/pisces/assets/pisces-1080p.png "Pisces screenshot 1")
![alt text](https://gamesfreaksa.info/pisces/assets/pisces-1080p-2.png "Pisces screenshot 2")

## System Minimum Requirements
- OS: Windows 10 64-bit
- Memory: 4GB
- Resolution: 1920 x 1080

## Installation
Download the zip file, unzip it whereever you want, and run Pisces.exe.

## Camera Mirroring
If you want to use the camera mirroring feature (Turned on/off in the `O`ptions screen), you'll need to install [OBS](https://obsproject.com/download) version 28 or higher. Currently we use its Virtual Camera to render the mirrored view that doesn't display hidden zone cards.

# Keybinds
|Function|Inputs|
|-|-|
|Select/drag cards|Left Mouse Button (LMB)|
|Open context menus|Right Mouse Button (RMB)|
|&nbsp;&nbsp;(on background or on a card)||

To perform an action on many cards (if supported, such as tapping or sending to another zone), you can explicitly select cards with rubber-band selection over them (pressing LMB on the background and dragging a selection rectangle). You can also add cards to the selection by holding `Ctrl` when you LMB click on a card or while doing a rubber-band selection.

But because Pisces makes it easy to do these actions on a single card that you've simply hovered the cursor over, when it says "...cards" below with an action, it means that it will apply to the cards that are explicitly selected *and* the card that is hovered (or right-click context menu was opened on) if both situations are going on.

*Note: It is possible to select cards on the battlefield and then hover and start dragging a card in your Hand and they all start dragging, so watch out for things like that.*

Another thing is some of these have *N* in the action. What that means is you can type a number before hitting the keybind to give the action a parameter. They'll default to doing the action once if you don't, but typing `1`, `3`, `D` will cause you to draw 13 cards (if there are at least 13 left in the Deck). *Currently the entered number will be clamped to no more than 100 before being used.*

## Manipulating cards
|Action|Inputs|
|-|-|
|Tap cards|`T`|
|Flip cards (front/back)|`F`|
|Flip cards (upside-down)|`Shift`+`F`|
|Duplicate cards|`Z`|
|Add *N* counters to cards|`Numpad +` or `=`/`+` key|
|Remove *N* counters from cards|`Numpad -` or `-`/`_` key|

## Card Zone Changes
|Action|Inputs|
|-|-|
|Delete cards (they're *gone*)|`X`|
|Send cards to Hand|`H`|
|Send cards to top of Deck|`J`|
|Send cards to bottom of Deck|`K`|
|Send cards to Graveyard|`G`|
|Send cards to Exile|`E`|
|Send cards to Command Zone|`C`|

## Zone Functions
|Action|Inputs|
|-|-|
|Open Deck|`F2`|
|Open Graveyard|`F3`|
|Open Exile|`F4`|
|Open Command Zone|`F5`|
|Shuffle Deck|`S`|
|Scry *N* cards|`R`|
|Draw *N* cards|`D`|
|Mill *N* cards|`M`|

Scry is an additional view that begins when you hit `R` to add 1 or more cards to it. If you just hit `R` it will add one card from the top of the Deck, but you can repeat `R` to add more cards from the top of the deck, or enter a number before hitting `R` to put that many cards in the Scry view.

It exits then you hit `Esc` or no cards are left in it. By default, any cards remaining in Scry when you use `Esc` are put on the bottom of the Deck in a random order. Some of the blow modify `Esc` to send them somewhere else.

### Scry Keybinds
|Action|Inputs|
|-|-|
|Put remaining into Graveyard|`Shift`+`Esc`|
|Put remaining into Exile|`Tab`+`Esc`|
|Toggle Reveal mode|`Ctrl`|
|&nbsp;&nbsp;(Opponents see them via Camera Mirror)||

## Other
|Action|Inputs|
|-|-|
|Untap all|`N`|
|Open Options|`O`|
|Save state|`F7`|
|Load state|`F8`|
|Coalesce selected cards being dragged|`Q`|

The Coalesce action deserves its own explanation. It will only happen *while* you are dragging cards that are all explicitly selected &ndash; not dragging one that you only hovered &ndash; and hit `Q`. It will form those cards into a neat diagonal arrangement by default.

Each time you hit `Q` while still in dragging mode, the frontmost card of the group will be sent to the back, allowing you to cycle through them. To help you better control this, be aware that the card your cursor is over when you begin the dragging is the first one to the back when you first hit `Q`.

As rudimentary ways to do vertical and horizontal arrangements, you can do `Shift`+`Q` to make it cycle front to back and arrange them vertically, or the same but horizontally with `Ctrl`+`Q`.

# Feedback and Bug Reports
To give us feedback, suggestions, to get help, or to submit a bug report, please visit the [Discord server](https://discord.gg/XxrzyphTPd).


CHANGELOG.md
# 2023-02-03
- **The F-key shortcuts have been changed to allow F1 to open a help screen in the future.** So now `F2` for Deck, `F3` for Graveyard, `F4` for Exile, and `F5` for Command Zone.
- Added upside-down flip with `Shift`+`F`. (No right-click menu entry yet)
- Fixed a bug where only the keyboard top row numbers could be used to preceed `D`, `M`, or `R` keybinds with an amount of cards to Draw, Mill, or Scry. Now the numpad numbers work, too. (Also the add/remove counters keybinds, below)
- Fixed issue where using `E` (exile), `C` (command zone), or `K` (bottom of deck) on a card that was already in the that zone caused an error and crash.
- Refactored counter add/remove keybinds &ndash; the `Numpad +` and `Numpad -` keys, minus/underscore key, and equal/plus key (but you don't have to hold shift) &ndash; to use the repeatable event like Draw, Mill, and Scry do so you can preceed it with numbers to add that many counters. Example: type `1`, `2`, and then `+` to add counters, and it'll add 12 counters.
- Allow Flip (front/back) of cards that are selected, not just the card that's hovered. In case of werewolves... or Ixidron.
- Holding `Ctrl` before starting the rubber-band selection will add cards to the current selection instead of starting over.
- `Ctrl` + left click on a card adds it to the current selection.
- Adjusted `gesture_drag_time` and `gesture_double_tap_time` so that double-tapping a card on the battlefield successfully toggles whether it's tapped or untapped.
- Added options for auto-deselect after tapping cards, and auto-deselecting after dragging and dropping them.
- Added simple ini settings that save the two auto-deselect options and the "Draw a card when starting a new turn" option (the `N` key that untaps everything - doesn't actually track turns).
- The camera mirroring button in Options now tells you if clicking it will turn it on or off.
- Save State should now track some other card attributes like flipped, upside-down, tapped, number of counters, and annotation.
- These are still rudimentary, but normal use of `Q` to coalesce arranges them diagonally - what's new is holding `Shift` to arrange them vertically or holding `Ctrl` to arrange them horizontally. It will still cycle top card to the back.
- Mostly, when cards are moved off the battlefield to one of the other zones, their statuses for being face-down or upside-down are reset to upright and front-face showing, and all cards in Graveyard and Exile are visible to Camera Mirroring (Hand and Deck being entirely hidden to the camera). There is now an exception for this if you hold `Shift` when exiling a face-down card with `E` or if you are holding `Shift` when you drop a face-down card into the Exile stack. That card becomes hidden to the Camera Mirroring and flipping it face-up in Exile will allow you to see it without opponents seeing it, so that you can exile cards face-down somewhat properly. Hovering that card in Exile and using `E` or doing a brief drag of the card and dropping it into Exile again _without_ holding `Shift` will reset it to face-up and be visible to opponents.

---
# 2022-12-17
- Fixed bugs with loading bad background/sleeve images causing a crash.
- Added an error window for when loading a background or sleeve image fails.
- Fixed bugs where cards put into the graveyard with Mill (`M` keybind) or from modifying `Esc` in the Scry view would put them at the bottom instead of the top.
- Fixed a bug where dragging card(s) to a hidden zone (Deck or Hand) was not clearing the counters.
- Added keybind hints to context menus.
- Some other misc. bugfixes.

mmap_buffer_gml/mmap_buffer.cpp
#include <windows.h>
#include <stdio.h>
#include <conio.h>
#include <tchar.h>
#include <vector>

using namespace std;

// File Pointers
vector<HANDLE> handles;
vector<LPCTSTR> files;

extern "C" __declspec(dllexport) double open_file(char* tag, double size)
{
    // Convert the string to LPCWSTR
    WCHAR szName[256];
    memset(szName, 0, sizeof(szName));
    MultiByteToWideChar(CP_ACP, 0, tag, strlen(tag) + 1, szName,
        sizeof(szName) / sizeof(szName[0]));

    int pageSize = (int)size;

    HANDLE hMapFile = CreateFileMapping(
        INVALID_HANDLE_VALUE,    // use paging file
        NULL,                    // default security
        PAGE_READWRITE,          // read/write access
        0,                       // maximum object size (high-order DWORD)
        pageSize,                // maximum object size (low-order DWORD)
        szName);                 // name of mapping object

    if (hMapFile == NULL)
    {
        _tprintf(TEXT("Could not create file mapping object (%d).\n"),
            GetLastError());
        return 1;
    }

    LPCTSTR pBuf = (LPTSTR)MapViewOfFile(hMapFile,   // handle to map object
        FILE_MAP_ALL_ACCESS, // read/write permission
        0,
        0,
        pageSize);

    if (pBuf == NULL)
    {
        _tprintf(TEXT("Could not map view of file (%d).\n"),
            GetLastError());

        CloseHandle(hMapFile);

        return 1;
    }

    handles.push_back(hMapFile);
    files.push_back(pBuf);
    return handles.size() - 1;
}

extern "C" __declspec(dllexport) double close_file(double index)
{
    LPCTSTR pBuf = files[index];
    HANDLE hMapFile = handles[index];

    if (pBuf != NULL) {
        UnmapViewOfFile(pBuf);
        files[index] = NULL;
    }

    if (hMapFile != NULL)
    {
        CloseHandle(hMapFile);
        handles[index] = NULL;
    }
    
    return 0;
}

extern "C" __declspec(dllexport) double copy_to_file(double index, char* data, double size, double offset)
{
    char* pBuf = (char*) files[index];
    pBuf += (int)offset;

    if (pBuf != NULL)
    {
        CopyMemory((PVOID) pBuf, data, size);
    }

    return 0;
}

extern "C" __declspec(dllexport) double copy_from_file(double index, char* data, double size)
{
    LPCTSTR pBuf = files[index];

    if (pBuf != NULL)
    {
        CopyMemory(data, (PVOID)pBuf, size);
    }

    return 0;
}

mmap_buffer_gml/mmap_buffer_gml/image_formats.h
#pragma once

#include <cmath>
#include <libyuv.h>

// libyuv names RGBA-type formats after the order in a *register*,
// whereas we name it after the order in *memory*.
// For example, libyuv ARGB is referred to as BGRA in function names below.

// Passing a negative height to the functions below flips the image vertically.

// copy
static void gray_to_bgra(const uint8_t* gray, uint8_t* bgra, int32_t width, int32_t height) {
    libyuv::J400ToARGB(
        gray, width,
        bgra, width * 4,
        width, height);
}

// copy
static void rgb_to_bgra(const uint8_t* rgb, uint8_t* bgra, int32_t width, int32_t height) {
    libyuv::RAWToARGB(
        rgb, width * 3,
        bgra, width * 4,
        width, height);
}

// copy
static void bgra_to_rgba(const uint8_t* bgra, uint8_t* rgba, int32_t width, int32_t height) {
    libyuv::ARGBToABGR(
        bgra, width * 4,
        rgba, width * 4,
        width, height);
}

// copy
static void bgra_to_bgra(const uint8_t* src, uint8_t* dist, int32_t width, int32_t height) {
    libyuv::ARGBCopy(
        src, width * 4,
        dist, width * 4,
        width, height);
}

#define rgba_to_rgba bgra_to_bgra

// horizontal and vertical subsampling and yuv conversion
static void rgb_to_i420(const uint8_t* rgb, uint8_t* i420, int32_t width, int32_t height) {
    int32_t height_ = height;
    height = std::abs(height);
    int32_t half_width = width / 2;
    int32_t half_height = height / 2;

    libyuv::RAWToI420(
        rgb, width * 3,
        i420, width,
        i420 + width * height, half_width,
        i420 + width * height + half_width * half_height, half_width,
        width, height_);
}

// copy
static void bgr_to_bgra(const uint8_t* bgr, uint8_t* bgra, int32_t width, int32_t height) {
    libyuv::RGB24ToARGB(
        bgr, width * 3,
        bgra, width * 4,
        width, height);
}

// horizontal and vertical subsampling and yuv conversion
static void bgr_to_i420(const uint8_t* bgr, uint8_t* i420, int32_t width, int32_t height) {
    int32_t height_ = height;
    height = std::abs(height);
    int32_t half_width = width / 2;
    int32_t half_height = height / 2;

    libyuv::RGB24ToI420(
        bgr, width * 3,
        i420, width,
        i420 + width * height, half_width,
        i420 + width * height + half_width * half_height, half_width,
        width, height_);
}

// horizontal and vertical subsampling and yuv conversion
static void bgra_to_nv12(const uint8_t* bgra, uint8_t* nv12, int32_t width, int32_t height) {
    int32_t height_ = height;
    height = std::abs(height);
    libyuv::ARGBToNV12(
        bgra, width * 4,
        nv12, width,
        nv12 + width * height, width,
        width, height_);
}

static void abgr_to_i420(const uint8_t* abgr, uint8_t* i420, int32_t width, int32_t height) {
    int32_t height_ = height;
    height = std::abs(height);
    int32_t half_width = width / 2;
    int32_t half_height = height / 2;

    libyuv::ABGRToI420(
        abgr, width * 4,
        i420, width,
        i420 + width * height, half_width,
        i420 + width * height + half_width * half_height, half_width,
        width, height_);
}

// horizontal subsampling and yuv conversion
static void bgra_to_uyvy(const uint8_t* bgra, uint8_t* uyvy, int32_t width, int32_t height) {
    libyuv::ARGBToUYVY(
        bgra, width * 4,
        uyvy, width * 2,
        width, height);
}

// copy
static void i420_to_nv12(const uint8_t* i420, uint8_t* nv12, int32_t width, int32_t height) {
    int32_t height_ = height;
    height = std::abs(height);
    int32_t half_width = width / 2;
    int32_t half_height = height / 2;

    libyuv::I420ToNV12(
        i420, width,
        i420 + width * height, half_width,
        i420 + width * height + half_width * half_height, half_width,
        nv12, width,
        nv12 + width * height, width,
        width, height_);
}

// horizontal and vertical upsampling and yuv conversion
static void i420_to_bgra(const uint8_t* i420, uint8_t* bgra, int32_t width, int32_t height) {
    int32_t height_ = height;
    height = std::abs(height);
    int32_t half_width = width / 2;
    int32_t half_height = height / 2;

    libyuv::I420ToARGB(
        i420, width,
        i420 + width * height, half_width,
        i420 + width * height + half_width * half_height, half_width,
        bgra, width * 4,
        width, height_);
}

// horizontal and vertical upsampling and yuv conversion
static void i420_to_rgba(const uint8_t* i420, uint8_t* rgba, int32_t width, int32_t height) {
    int32_t height_ = height;
    height = std::abs(height);
    int32_t half_width = width / 2;
    int32_t half_height = height / 2;

    libyuv::I420ToABGR(
        i420, width,
        i420 + width * height, half_width,
        i420 + width * height + half_width * half_height, half_width,
        rgba, width * 4,
        width, height_);
}

// copy
static void nv12_to_i420(const uint8_t* nv12, uint8_t* i420, int32_t width, int32_t height) {
    int32_t height_ = height;
    height = std::abs(height);
    int32_t half_width = width / 2;
    int32_t half_height = height / 2;

    libyuv::NV12ToI420(
        nv12, width,
        nv12 + width * height, width,
        i420, width,
        i420 + width * height, half_width,
        i420 + width * height + half_width * half_height, half_width,
        width, height_);
}

// horizontal and vertical upsampling and yuv conversion
static void nv12_to_bgra(const uint8_t* nv12, uint8_t* bgra, int32_t width, int32_t height) {
    int32_t height_ = height;
    height = std::abs(height);
    libyuv::NV12ToARGB(
        nv12, width,
        nv12 + width * height, width,
        bgra, width * 4,
        width, height_);
}

// horizontal and vertical upsampling and yuv conversion
static void nv12_to_rgba(const uint8_t* nv12, uint8_t* rgba, int32_t width, int32_t height) {
    int32_t height_ = height;
    height = std::abs(height);
    libyuv::NV12ToABGR(
        nv12, width,
        nv12 + width * height, width,
        rgba, width * 4,
        width, height_);
}

// vertical upsampling
static void i420_to_uyvy(const uint8_t* i420, uint8_t* uyvy, int32_t width, int32_t height) {
    int32_t height_ = height;
    height = std::abs(height);
    int32_t half_width = width / 2;
    int32_t half_height = height / 2;

    libyuv::I420ToUYVY(
        i420, width,
        i420 + width * height, half_width,
        i420 + width * height + half_width * half_height, half_width,
        uyvy, width * 2,
        width, height_);
}

// vertical subsampling
static void yuyv_to_nv12(const uint8_t* yuyv, uint8_t* nv12, int32_t width, int32_t height) {
    int32_t height_ = height;
    height = std::abs(height);
    libyuv::YUY2ToNV12(
        yuyv, width * 2,
        nv12, width,
        nv12 + width * height, width,
        width, height_);
}

// vertical subsampling
static void yuyv_to_i420(const uint8_t* yuyv, uint8_t* i420, int32_t width, int32_t height) {
    int32_t height_ = height;
    height = std::abs(height);
    int32_t half_width = width / 2;
    int32_t half_height = height / 2;

    libyuv::YUY2ToI420(
        yuyv, width * 2,
        i420, width,
        i420 + width * height, half_width,
        i420 + width * height + half_width * half_height, half_width,
        width, height_);
}

// copy
static void yuyv_to_i422(const uint8_t* yuyv, uint8_t* i422, int32_t width, int32_t height) {
    int32_t height_ = height;
    height = std::abs(height);
    int32_t half_width = width / 2;

    libyuv::YUY2ToI422(
        yuyv, width * 2,
        i422, width,
        i422 + width * height, half_width,
        i422 + width * height + half_width * height, half_width,
        width, height_);
}

// horizontal upsampling and yuv conversion
static void yuyv_to_bgra(const uint8_t* yuyv, uint8_t* bgra, int32_t width, int32_t height) {
    libyuv::YUY2ToARGB(
        yuyv, width * 2,
        bgra, width * 4,
        width, height);
}

// vertical subsampling
static void uyvy_to_nv12(const uint8_t* uyvy, uint8_t* nv12, int32_t width, int32_t height) {
    int32_t height_ = height;
    height = std::abs(height);
    libyuv::UYVYToNV12(
        uyvy, width * 2,
        nv12, width,
        nv12 + width * height, width,
        width, height_);
}

// copy
static void i422_to_uyvy(const uint8_t* i422, uint8_t* uyvy, int32_t width, int32_t height) {
    int32_t height_ = height;
    height = std::abs(height);
    int32_t half_width = width / 2;

    libyuv::I422ToUYVY(
        i422, width,
        i422 + width * height, half_width,
        i422 + width * height + half_width * height, half_width,
        uyvy, width * 2,
        width, height_);
}

// horizontal upsampling and yuv conversion
static void uyvy_to_bgra(const uint8_t* uyvy, uint8_t* bgra, int32_t width, int32_t height) {
    libyuv::UYVYToARGB(
        uyvy, width * 2,
        bgra, width * 4,
        width, height);
}

static int32_t bgra_frame_size(int32_t width, int32_t height) {
    return width * height * 4;
}

static int32_t gray_frame_size(int32_t width, int32_t height) {
    return width * height;
}

static int32_t i420_frame_size(int32_t width, int32_t height) {
    return width * height * 3 / 2;
}

static int32_t i422_frame_size(int32_t width, int32_t height) {
    return width * height * 2;
}

#define rgba_frame_size bgra_frame_size
#define nv12_frame_size i420_frame_size
#define uyvy_frame_size i422_frame_size
#define yuyv_frame_size i422_frame_size

mmap_buffer_gml/mmap_buffer_gml/virtual_output.h
#pragma once

#include <stdio.h>
#include <Windows.h>
#include <vector>
#include <optional>
#include "shared-memory-queue.h"
#include "image_formats.h"

class VirtualOutput {
private:
    bool _output_running = false;
    video_queue_t* _vq;
    uint32_t _frame_width;
    uint32_t _frame_height;
    uint32_t _frame_fourcc;
    std::vector<uint8_t> _buffer_tmp;
    std::vector<uint8_t> _buffer_output;
    bool _have_clockfreq = false;
    LARGE_INTEGER _clock_freq;

    uint64_t get_timestamp_ns()
    {
        if (!_have_clockfreq) {
            QueryPerformanceFrequency(&_clock_freq);
            _have_clockfreq = true;
        }

        LARGE_INTEGER current_time;
        QueryPerformanceCounter(&current_time);
        double time_val = (double)current_time.QuadPart;
        time_val *= 1000000000.0;
        time_val /= (double)_clock_freq.QuadPart;

        return static_cast<uint64_t>(time_val);
    }

public:
    VirtualOutput(uint32_t width, uint32_t height, double fps, uint32_t fourcc) {
        // https://github.com/obsproject/obs-studio/blob/9da6fc67/.github/workflows/main.yml#L484
        LPCWSTR guid = L"CLSID\\{A3FCE0F5-3493-419F-958A-ABA1250EC20B}";
        HKEY key = nullptr;
        if (RegOpenKeyExW(HKEY_CLASSES_ROOT, guid, 0, KEY_READ, &key) != ERROR_SUCCESS) {
            throw std::runtime_error(
                "OBS Virtual Camera device not found! "
                "Did you install OBS?"
            );
        }

        _frame_fourcc = libyuv::CanonicalFourCC(fourcc);
        _frame_width = width;
        _frame_height = height;

        uint32_t out_frame_size = nv12_frame_size(width, height);

        switch (_frame_fourcc) {
        case libyuv::FOURCC_ABGR:
        case libyuv::FOURCC_RAW:
        case libyuv::FOURCC_24BG:
            // RGB|BGR|RGBA -> I420 -> NV12
            _buffer_tmp.resize(i420_frame_size(width, height));
            _buffer_output.resize(out_frame_size);
            break;
        case libyuv::FOURCC_J400:
            // GRAY -> BGRA -> NV12
            _buffer_tmp.resize(bgra_frame_size(width, height));
            _buffer_output.resize(out_frame_size);
            break;
        case libyuv::FOURCC_I420:
        case libyuv::FOURCC_YUY2:
        case libyuv::FOURCC_UYVY:
            // I420|YUYV|UYVY -> NV12
            _buffer_output.resize(out_frame_size);
            break;
        case libyuv::FOURCC_NV12:
            break;
        default:
            throw std::runtime_error(
                "Unsupported image format."
            );
        }

        uint64_t interval = (uint64_t)(10000000.0 / fps);

        _vq = video_queue_create(width, height, interval);

        if (!_vq) {
            throw std::runtime_error("virtual camera output could not be started");
        }

        _output_running = true;
    }

    void stop()
    {
        if (!_output_running) {
            return;
        }
        video_queue_close(_vq);
        _output_running = false;
    }

    void send(const uint8_t* frame)
    {
        if (!_output_running)
            return;

        uint8_t* tmp = _buffer_tmp.data();
        uint8_t* out_frame;

        switch (_frame_fourcc) {
        case libyuv::FOURCC_RAW:
            out_frame = _buffer_output.data();
            rgb_to_i420(frame, tmp, _frame_width, _frame_height);
            i420_to_nv12(tmp, out_frame, _frame_width, _frame_height);
            break;
        case libyuv::FOURCC_24BG:
            out_frame = _buffer_output.data();
            bgr_to_i420(frame, tmp, _frame_width, _frame_height);
            i420_to_nv12(tmp, out_frame, _frame_width, _frame_height);
            break;
        case libyuv::FOURCC_ABGR:
            out_frame = _buffer_output.data();
            abgr_to_i420(frame, tmp, _frame_width, _frame_height);
            i420_to_nv12(tmp, out_frame, _frame_width, _frame_height);
            break;
        case libyuv::FOURCC_J400:
            out_frame = _buffer_output.data();
            gray_to_bgra(frame, tmp, _frame_width, _frame_height);
            bgra_to_nv12(tmp, out_frame, _frame_width, _frame_height);
            break;
        case libyuv::FOURCC_I420:
            out_frame = _buffer_output.data();
            i420_to_nv12(frame, out_frame, _frame_width, _frame_height);
            break;
        case libyuv::FOURCC_NV12:
            out_frame = const_cast<uint8_t*>(frame);
            break;
        case libyuv::FOURCC_YUY2:
            out_frame = _buffer_output.data();
            yuyv_to_nv12(frame, out_frame, _frame_width, _frame_height);
            break;
        case libyuv::FOURCC_UYVY:
            out_frame = _buffer_output.data();
            uyvy_to_nv12(frame, out_frame, _frame_width, _frame_height);
            break;
        default:
            throw std::logic_error("not implemented");
        }

        // NV12 has two planes
        uint8_t* y = out_frame;
        uint8_t* uv = out_frame + _frame_width * _frame_height;

        // One entry per plane
        uint32_t linesize[2] = { _frame_width, _frame_width / 2 };
        uint8_t* data[2] = { y, uv };

        uint64_t timestamp = get_timestamp_ns();

        video_queue_write(_vq, data, linesize, timestamp);
    }

    std::string device()
    {
        // https://github.com/obsproject/obs-studio/blob/eb98505a2/plugins/win-dshow/virtualcam-module/virtualcam-module.cpp#L196
        return "OBS Virtual Camera";
    }

    uint32_t native_fourcc() {
        return libyuv::FOURCC_NV12;
    }
};

mmap_buffer_gml/mmap_buffer_gml/shared-memory-queue.cpp
#include <windows.h>
#include "shared-memory-queue.h"

#define VIDEO_NAME L"OBSVirtualCamVideo"

enum queue_type {
	SHARED_QUEUE_TYPE_VIDEO,
};

struct queue_header {
	volatile uint32_t write_idx;
	volatile uint32_t read_idx;
	volatile uint32_t state;

	uint32_t offsets[3];

	uint32_t type;

	uint32_t cx;
	uint32_t cy;
	uint64_t interval;

	uint32_t reserved[8];
};

struct video_queue {
	HANDLE handle;
	bool ready_to_read;
	struct queue_header* header;
	uint64_t* ts[3];
	uint8_t* frame[3];
	long last_inc;
	int dup_counter;
	bool is_writer;
};

#define ALIGN_SIZE(size, align) size = (((size) + (align - 1)) & (~(align - 1)))
#define FRAME_HEADER_SIZE 32

video_queue_t* video_queue_create(uint32_t cx, uint32_t cy, uint64_t interval)
{
	struct video_queue vq = { 0 };
	struct video_queue* pvq;
	DWORD frame_size = cx * cy * 3 / 2;
	uint32_t offset_frame[3];
	DWORD size;

	size = sizeof(struct queue_header);

	ALIGN_SIZE(size, 32);

	offset_frame[0] = size;
	size += frame_size + FRAME_HEADER_SIZE;
	ALIGN_SIZE(size, 32);

	offset_frame[1] = size;
	size += frame_size + FRAME_HEADER_SIZE;
	ALIGN_SIZE(size, 32);

	offset_frame[2] = size;
	size += frame_size + FRAME_HEADER_SIZE;
	ALIGN_SIZE(size, 32);

	struct queue_header header = { 0 };

	header.state = SHARED_QUEUE_STATE_STARTING;
	header.cx = cx;
	header.cy = cy;
	header.interval = interval;
	vq.is_writer = true;

	for (size_t i = 0; i < 3; i++) {
		uint32_t off = offset_frame[i];
		header.offsets[i] = off;
	}

	/* fail if already in use */
	vq.handle = OpenFileMappingW(FILE_MAP_READ, false, VIDEO_NAME);
	if (vq.handle) {
		CloseHandle(vq.handle);
		return NULL;
	}

	vq.handle = CreateFileMappingW(INVALID_HANDLE_VALUE, NULL,
		PAGE_READWRITE, 0, size, VIDEO_NAME);
	if (!vq.handle) {
		return NULL;
	}

	vq.header = (struct queue_header*)MapViewOfFile(
		vq.handle, FILE_MAP_ALL_ACCESS, 0, 0, 0);
	if (!vq.header) {
		CloseHandle(vq.handle);
		return NULL;
	}
	memcpy(vq.header, &header, sizeof(header));

	for (size_t i = 0; i < 3; i++) {
		uint32_t off = offset_frame[i];
		vq.ts[i] = (uint64_t*)(((uint8_t*)vq.header) + off);
		vq.frame[i] = ((uint8_t*)vq.header) + off + FRAME_HEADER_SIZE;
	}
	pvq = (video_queue*) malloc(sizeof(vq));
	if (!pvq) {
		CloseHandle(vq.handle);
		return NULL;
	}
	memcpy(pvq, &vq, sizeof(vq));
	return pvq;
}

video_queue_t* video_queue_open()
{
	struct video_queue vq = { 0 };

	vq.handle = OpenFileMappingW(FILE_MAP_READ, false, VIDEO_NAME);
	if (!vq.handle) {
		return NULL;
	}

	vq.header = (struct queue_header*)MapViewOfFile(
		vq.handle, FILE_MAP_READ, 0, 0, 0);
	if (!vq.header) {
		CloseHandle(vq.handle);
		return NULL;
	}

	struct video_queue* pvq = (video_queue *) malloc(sizeof(vq));
	if (!pvq) {
		CloseHandle(vq.handle);
		return NULL;
	}
	memcpy(pvq, &vq, sizeof(vq));
	return pvq;
}

void video_queue_close(video_queue_t* vq)
{
	if (!vq) {
		return;
	}
	if (vq->is_writer) {
		vq->header->state = SHARED_QUEUE_STATE_STOPPING;
	}

	UnmapViewOfFile(vq->header);
	CloseHandle(vq->handle);
	free(vq);
}

void video_queue_get_info(video_queue_t* vq, uint32_t* cx, uint32_t* cy,
	uint64_t* interval)
{
	struct queue_header* qh = vq->header;
	*cx = qh->cx;
	*cy = qh->cy;
	*interval = qh->interval;
}

#define get_idx(inc) ((unsigned long)inc % 3)

void video_queue_write(video_queue_t* vq, uint8_t** data, uint32_t* linesize,
	uint64_t timestamp)
{
	struct queue_header* qh = vq->header;
	long inc = ++qh->write_idx;

	unsigned long idx = get_idx(inc);
	size_t size = linesize[0] * qh->cy;

	*vq->ts[idx] = timestamp;
	memcpy(vq->frame[idx], data[0], size);
	memcpy(vq->frame[idx] + size, data[1], size / 2);

	qh->read_idx = inc;
	qh->state = SHARED_QUEUE_STATE_READY;
}

enum queue_state video_queue_state(video_queue_t* vq)
{
	if (!vq) {
		return SHARED_QUEUE_STATE_INVALID;
	}

	enum queue_state state = (enum queue_state)vq->header->state;
	if (!vq->ready_to_read && state == SHARED_QUEUE_STATE_READY) {
		for (size_t i = 0; i < 3; i++) {
			size_t off = vq->header->offsets[i];
			vq->ts[i] = (uint64_t*)(((uint8_t*)vq->header) + off);
			vq->frame[i] = ((uint8_t*)vq->header) + off +
				FRAME_HEADER_SIZE;
		}
		vq->ready_to_read = true;
	}

	return state;
}

mmap_buffer_gml/mmap_buffer_gml/camera_send.cpp
#include <conio.h>
#include <stdio.h>
#include <tchar.h>
#include <windows.h>

#include <chrono>
#include <iostream>
#include <mutex>
#include <thread>
#include <vector>

#include "camera_send.h"
#include "virtual_output.h"


// Camera
VirtualOutput* camera;

// Buffer data from GMS
unsigned char* data_buff;
unsigned char* data_ready;
int data_buffs;

// Threads
std::thread thread_thing;
std::atomic<bool> shouldQuit;
std::mutex mut;

extern "C" __declspec(dllexport) double camera_is_installed()
{
    LPCWSTR guid = L"CLSID\\{A3FCE0F5-3493-419F-958A-ABA1250EC20B}";
    HKEY key = nullptr;

    if (RegOpenKeyExW(HKEY_CLASSES_ROOT, guid, 0, KEY_READ, &key) != ERROR_SUCCESS) {
        return 0;
    }
    else
    {
        RegCloseKey(key);
        return 1;
    }
}

extern "C" __declspec(dllexport) double start_camera(unsigned char* buff, unsigned char* ready, double buff_nums)
{
    std::lock_guard<std::mutex> lock(mut);

    if (camera != NULL)
    {
        return 0;
    }

    int width = 1920;
    int height = 1080;
    int fps = 60;

    data_buff = buff;
    data_ready = ready;
    data_buffs = (int) buff_nums;

    camera = new VirtualOutput(width, height, fps, libyuv::FOURCC_ABGR);
    thread_thing = std::thread(camera_impl);

    shouldQuit = false;

    return 0;
}

extern "C" __declspec(dllexport) double stop_camera()
{
    std::lock_guard<std::mutex> lock(mut);

    if (camera != NULL)
    {
        shouldQuit = true;
        thread_thing.join();
        camera->stop();
        delete camera;
        camera = NULL;
    }

    return 0;
}

void camera_impl()
{
    int curr_buff = 0;
    while (TRUE)
    {
        std::this_thread::sleep_for(std::chrono::microseconds(50));

        if (!shouldQuit)
        {
            // Send frame
            unsigned char* buff = &data_buff[curr_buff * 1920 * 1080 * 4];
            unsigned char* ready = &data_ready[curr_buff];

            if (*ready == 2)
            {
                camera->send(buff);
                *ready = 0;
                curr_buff = (curr_buff + 1) % data_buffs;
            }
        }
        else
        {
            return;
        }
    }
}
